### ISSUE TITLE
Add gpsro data to the conventional data monitor

### ISSUE THREAD

**Issue by EdwardSafford-NOAA** [NOAA-EMC/GSI#62]
Add support for gpsro data into the existing format of the Conventional data monitor (ConMon).  This is a reissue of Redmine issue [51902](https://vlab.ncep.noaa.gov/redmine/issues/51902).

**Comment by EdwardSafford-NOAA** [comment]
For reference, this from Kristen:

```
1. station_id - example 02690031
0269 is the satellite identifier, e.g., 269 is sprite, 750 is COSMIC2 E1.
0031 is the gnss reference satellite id

2. gps_type
gps_type=1 means bending angle is assimilated, and 0 means refractivity is assimilated.

3. observation_subtype
observation_subtype makes more sense for conventional data. For GNSSRO data, this variable is used to store the profile id.  
```

**Comment by EdwardSafford-NOAA** [comment]
More useful info from Kristen:

```
1. We only assimilate bending angle observations, but I think refractivity is treated the same in the diag files. You probably don't need to worry about refractivity.

2. The pressure field is actually the pressure, and the height is the impact height, which is basically the same thing as the altitude. When an RO profile is measured, observations are taken at different heights as the LEO moves behind the earth. So if you print out the heights, you will see that they increase to maybe 60 km. So one RO profile would be all of the observations from the lowest height (maybe 2 km) until the heights stop increasing. But not all profiles go from 2 km to 60 km, and not all profiles contain the same number of observations.

Vertical plots of OMG/G are usually helpful, like this one. My IDL generates this as
 gps(n).depart=data.obs_minus_forecast_adjusted(n)/(data.observation(n)-data.obs_minus_forecast_adjusted(n))
```

**Comment by EdwardSafford-NOAA** [comment]
Based on what I've learned so far it seems that in order to get the gps data to fit inside the existing ConMon framework is to write a data extraction executable unique to the gps diag files.  I don't see a good way to reuse any of the existing executables; the gps data is different enough that modifying an executable would be a mess.  Also Kristen's vertical plots are quite different from what's available in the ConMon plot scripts now.

That makes me wonder if a better approach is to try to directly access the gps diag files using python scripts, both for error checking/reporting and plotting.  That's the direction we need to move in for all the *Mon packages anyway, so it seems as good a time as any to give that a go.  I'm investigating that angle now.

**Comment by RussTreadon-NOAA** [comment]
Exploring the direct reading of GSI netcdf diagnostic files via python
scripts and plotting via python utilities is a very worthwhile avenue of
development.  Glad to hear you are investigating this angle.

**Comment by EdwardSafford-NOAA** [comment]
After some study of the existing ConMon organization and experimentation with the PyGSI package these are my conclusions and resulting plans:

1. The existing ConMon is organized by pressure.  Per Kristen height (altitude) has more meaning for RO data than pressure.  So to produce corresponding graphics for the GPS data within the existing ConMon package I would have to add a new extraction executable and new plotting scripts organized by pressure.  At this point writing another way to extract data from NetCDF files and creating new GrADS scripts doesn't make much sense in terms of direction of the organization and long-term utility.
2. The PyGSI package currently supports binning data by pressure levels.  It doesn't support binning by height, though adding that capability does not appear to be a difficult prospect.  I was easily able to add height to the Conventional class (in pyGSI/diag.py) in my experiments.  The existing get_data() function, which optionally arranges the data into input pressure bins, could be copied as get_data_height() and modified to returned data binned by height.
3. This seems a good opportunity to begin to convert the ConMon package to a python based solution.  The existing package (which is not yet part of the operational DA package) can continue as is while the gps solution can be used as a template for adapting the existing ConMon to the yaml/python world.  Once the gps piece is in place and working to stakeholder satisfaction, the other ConMon pieces can be converted (I'll create a separate issue for that work).
4. This is dependent upon PyGSI becoming available on the wcoss systems.  My understanding is that that work is in progress.

**Comment by EdwardSafford-NOAA** [comment]
I've been able to extend the PyGSI package to allow binning of data by height instead of pressure.  I was able to do that within the existing routines in pyGSI/diag.py rather than copying existing routines and making them height specific, so that's the good news.  I also added plot_vertical() and _create_vertical_plot() routines to pyGSI/plot_data.py and that's working too.  Using thoes changes I've been able to create some prototype plots of individual gps sources.  Kristen gave me an example plot she's created using IDL and I've been trying to reproduce that as a starting point.  
This is her plot:
![265geoomb](https://user-images.githubusercontent.com/62339196/123102174-c8decd80-d402-11eb-82d0-9fae89f5c32e.png)

And this is my most recent effort:
![2021060518_conv_gps_O-F_265_assimilated_vertical](https://user-images.githubusercontent.com/62339196/123102395-02173d80-d403-11eb-9904-d9b054c1e12a.png)

Note that Kristen's plot is cumulative spanning more than one cycle, so the counts won't match, but the plotted curves are generally similar.  The bias/depart lines don't show much similarity though they are within the same range, so that may be ok.  The RMSE is the problem.  Mine is significantly greater than hers.  I continue to study why that is so; I may have an error in my data handling and/or rmse calculation.

**Comment by EdwardSafford-NOAA** [comment]
So far I've been unable to find a way in the PyGSI package to make the rmse and bias/depart numbers make sense.  It's probably me.  I decided to modify the ConMon's time data extraction executable to handle gps data and see if I can get it to generate better stats.  Though the ConMon is geared to pressure binning, the time series plots don't use that so it wasn't a big effort to make the extension.  If this works I can easily add this piece to the existing ConMon and maybe figure out what I'm doing wrong with the PyGSI.

**Comment by EdwardSafford-NOAA** [comment]
I've modified the ConMon's time data extraction executable and it now handles gps data, binning it by height in 5km increments, plus one level for 0-60 (all).  The counts are good but the rmse and bias calculations are still clearly wrong.  So I'm going to keep this modification and generate only time count files for gps data and call that a start.  It's worth noting that counts are the only thing for which the ConMon issues warnings for so this is consistent.  I've switched my ConMon cron on wcoss_d to use this developmental branch (feature/es_ConMon_62) for the GFS/gdas data.

I've updated the html/javascript on the server to modify the time/level menu on the fly for gps selections (in the Obs Type menu).  There already was a mechanism in place to switch between the pressure levels for q data and all others, so adding a third set of levels unique to gps wasn't much of a stretch.  I'm working now on verifying which gps types actually have data and am adding them to the time/obs type menu.

After I have enough data (~30 days) to establish norms for each of the gps types I'll generate historic count information and add them to the ConMon's automated check.  

I don't now plan on doing any more than what I've described to the ConMon's time extraction executable.  My next move will return to the PyGSI and revisit creating vertical plots with valid rmse and bias data.

**Comment by EdwardSafford-NOAA** [comment]
I'm pleased to report that I have been able to generate time series obs count plots for the gps data types.  I've verified the assimilated, rejected vqc, and monitored numbers and they agree with the figures in the gsistat file for the corresponding cycle.  The bias and rms values in the "stas" plots don't line up yet but I now have an idea of what's amiss there and will run some test plots with the next cycle.  This has been accomplished using the existing ConMon time extraction executable.  I didn't want to go that way but I was kind at a loss as to what I was doing wrong with the PyGSI package.  Working through the processing in the ConMon, I now believe I understand what I need to make the PyGSI create the vertical plots I'm looking for and will return to that as soon as I verify the "stas" plots line up with the gsistat figures.

![image](https://user-images.githubusercontent.com/62339196/125672422-8d296ea6-2e50-448e-939c-5dd002cda802.png)

**Comment by EdwardSafford-NOAA** [comment]
Commit 2f1149a added the gps observational counts to the ConMon's nobs.${PDY} files.  Once I've got 30 days worth of data there I can generate historic base files for the gps data sources and turn on warning generation for them.  Related to this work is [#PR57](https://github.com/NOAA-EMC/PyGSI/pull/57#pullrequestreview-719139037), which is the start of adapting the PyGSI package to gps monitoring needs.

**Comment by EdwardSafford-NOAA** [comment]
While waiting for the 30 days worth of data to finish this issue (see comment above), I sampled the image plot space usage.  I knew it was too big, but it's really too big.  

Horiz plots cover 22 regions and 8 levels and total 2.2G/cycle.
Time plots cover 10 regions and 13 levels and total 49.4G/cycle.
Vert plots cover 10 regions, 13 levels and total 40G/cycle.

I'm going to address that space usage as part of this issue.

The Time an Vert regions are defined as:
Global
N. Hemisphere
S. Hemisphere
Tropics
USA
Canada
N. America
S. America
Europe
Asia

I'll begin by disabling Time and Vert plots for N. Hemisphere, S. Hemisphere, and Canada.  I'll leave all the logic in place so it can be turned back on if needed.  That should reduce space usage for those by 30%.
